src/main/java/com/getindata/connectors/http/internal/sink/httpclient/BatchRequestSubmitter.java

```diff
@@ -0,0 +1,150 @@
+import java.net.http.HttpClient.Version;
+import java.net.http.HttpRequest.BodyPublisher;
+import java.net.http.HttpRequest.BodyPublishers;
+import java.net.http.HttpRequest.Builder;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.CompletableFuture;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.flink.annotation.VisibleForTesting;
+
+import com.getindata.connectors.http.internal.config.HttpConnectorConfigConstants;
+import com.getindata.connectors.http.internal.sink.HttpSinkRequestEntry;
+
+/**
+ * This implementation groups received events in batches and submits each batch as individual HTTP
+ * requests. Batch is created based on batch size or based on HTTP method type.
+ */
+@Slf4j
+public class BatchRequestSubmitter extends AbstractRequestSubmitter {
+
+    private static final byte[] BATCH_START_BYTES = "[".getBytes(StandardCharsets.UTF_8);
+
+    private static final byte[] BATCH_END_BYTES = "]".getBytes(StandardCharsets.UTF_8);
+
+    private static final byte[] BATCH_ELEMENT_DELIM_BYTES = ",".getBytes(StandardCharsets.UTF_8);
+
+    private final int httpReqeustBatchSize;
+
+    public BatchRequestSubmitter(
+            Properties properties,
+            String[] headersAndValue,
+            HttpClient httpClient) {
+
+        super(properties, headersAndValue, httpClient);
+
+        this.httpReqeustBatchSize = Integer.parseInt(
+            properties.getProperty(HttpConnectorConfigConstants.SINK_HTTP_BATCH_REQUEST_SIZE)
+        );
+    }
+
+    @Override
+    public List<CompletableFuture<JavaNetHttpResponseWrapper>> submit(
+            String endpointUrl,
+            List<HttpSinkRequestEntry> requestsToSubmit) {
+
+        if (requestsToSubmit.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        var responseFutures = new ArrayList<CompletableFuture<JavaNetHttpResponseWrapper>>();
+
+        int counter = 0;
+        String previousReqeustMethod = requestsToSubmit.get(0).method;
+        List<HttpSinkRequestEntry> reqeustBatch = new ArrayList<>(httpReqeustBatchSize);
+        for (var entry : requestsToSubmit) {
+            if (!previousReqeustMethod.equalsIgnoreCase(entry.method)) {
+                // break batch and submit
+                responseFutures.add(sendBatch(endpointUrl, reqeustBatch));
+                reqeustBatch.clear();
+                // start a new batch for new HTTP method.
+                reqeustBatch.add(entry);
+            } else {
+                reqeustBatch.add(entry);
+                if (++counter % httpReqeustBatchSize == 0) {
+                    // batch is full, submit and start new batch.
+                    responseFutures.add(sendBatch(endpointUrl, reqeustBatch));
+                    reqeustBatch.clear();
+                }
+            }
```

Comment:
I think it could be simplified a bit:
```suggestion
            if (!previousReqeustMethod.equalsIgnoreCase(entry.method) || counter % httpReqeustBatchSize == 0) {
                // break batch and submit
                responseFutures.add(sendBatch(endpointUrl, reqeustBatch));
                reqeustBatch.clear();
            } else {
                reqeustBatch.add(entry);
                counter++;
            }
```
Of course, it would mean that we will be sending the batch one iteration step later, but it should not be a problem anyway.

And btw, what's about `counter`? Wouldn't `requestBatch.size() >= httpRequestBatchSize` be sufficient? And, also, current version of the code has a small bug — `counter` is incremented only in the `else` branch, however in the `if` we're calling `requestBatch.add(entry)` too.

Reply:
changed, please let me know wht do you think now.

Reply:
@swtwsk ^